File: backend/src/main/java/com/nexus/jobboard/application/service/JobService.java
package com.nexus.jobboard.application.service;

import com.nexus.jobboard.application.dto.request.JobCreateRequest;
import com.nexus.jobboard.application.dto.request.JobUpdateRequest;
import com.nexus.jobboard.application.dto.request.JobSearchRequest;
import com.nexus.jobboard.application.dto.response.JobResponse;
import com.nexus.jobboard.domain.model.JobType;
import com.nexus.jobboard.domain.model.ExperienceLevel;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

/**
 * Job service interface following DIP and SRP
 * - Depends on abstractions, not concrete implementations
 * - Single responsibility: Job-related business operations
 */
public interface JobService {
    
    /**
     * Create a new job posting
     */
    JobResponse createJob(JobCreateRequest request, Long employerId);
    
    /**
     * Update existing job posting
     */
    JobResponse updateJob(Long jobId, JobUpdateRequest request, Long employerId);
    
    /**
     * Get job by ID
     */
    Optional<JobResponse> getJobById(Long jobId);
    
    /**
     * Get all active jobs with pagination
     */
    Page<JobResponse> getAllActiveJobs(Pageable pageable);
    
    /**
     * Get jobs by company
     */
    Page<JobResponse> getJobsByCompany(Long companyId, Pageable pageable);
    
    /**
     * Get jobs by category
     */
    Page<JobResponse> getJobsByCategory(Long categoryId, Pageable pageable);
    
    /**
     * Get jobs posted by specific user
     */
    Page<JobResponse> getJobsByUser(Long userId, Pageable pageable);
    
    /**
     * Search jobs with full-text search
     */
    Page<JobResponse> searchJobs(String searchTerm, Pageable pageable);
    
    /**
     * Advanced job filtering
     */
    Page<JobResponse> getJobsByFilters(String location, JobType jobType, 
                                     ExperienceLevel experienceLevel,
                                     BigDecimal minSalary, BigDecimal maxSalary,
                                     Boolean isRemote, Long categoryId,
                                     Pageable pageable);
    
    /**
     * Find jobs by required skills
     */
    Page<JobResponse> getJobsBySkills(List<Long> skillIds, Pageable pageable);
    
    /**
     * Deactivate job posting
     */
    void deactivateJob(Long jobId, Long employerId);
    
    /**
     * Activate job posting
     */
    void activateJob(Long jobId, Long employerId);
    
    /**
     * Delete job posting (soft delete by deactivation)
     */
    void deleteJob(Long jobId, Long employerId);
    
    /**
     * Get active job count by company
     */
    Long getActiveJobCountByCompany(Long companyId);
    
    /**
     * Get active job count by category
     */
    Long getActiveJobCountByCategory(Long categoryId);
    
    /**
     * Get active job count by user
     */
    Long getActiveJobCountByUser(Long userId);
    
    /**
     * Process expired jobs (background task)
     */
    void processExpiredJobs();
    
    /**
     * Get recommended jobs for a user (AI-powered)
     */
    Page<JobResponse> getRecommendedJobs(Long userId, Pageable pageable);
}

File: frontend/src/services/jobService.ts
import { AxiosResponse } from 'axios';
import { apiService } from './api';
import { Job, JobSearchFilters, PaginatedResponse } from '../types';

/**
 * Job service following SRP
 * - Single responsibility: Handle job-related API calls
 */
export class JobService {
  private readonly basePath = '/jobs';

  async getAllJobs(page = 0, size = 20): Promise<AxiosResponse<PaginatedResponse<Job>>> {
    return apiService.get<PaginatedResponse<Job>>(`${this.basePath}?page=${page}&size=${size}`);
  }

  async getJobById(id: number): Promise<AxiosResponse<Job>> {
    return apiService.get<Job>(`${this.basePath}/${id}`);
  }

  async searchJobs(
    filters: JobSearchFilters,
    page = 0,
    size = 20
  ): Promise<AxiosResponse<PaginatedResponse<Job>>> {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());

    if (filters.query) params.append('query', filters.query);
    if (filters.location) params.append('location', filters.location);
    if (filters.jobType) params.append('jobType', filters.jobType);
    if (filters.experienceLevel) params.append('experienceLevel', filters.experienceLevel);
    if (filters.minSalary) params.append('minSalary', filters.minSalary.toString());
    if (filters.maxSalary) params.append('maxSalary', filters.maxSalary.toString());
    if (filters.isRemote !== undefined) params.append('isRemote', filters.isRemote.toString());
    if (filters.categoryId) params.append('categoryId', filters.categoryId.toString());

    const endpoint = filters.query 
      ? `${this.basePath}/search?${params.toString()}`
      : `${this.basePath}/filter?${params.toString()}`;

    return apiService.get<PaginatedResponse<Job>>(endpoint);
  }

  async getJobsByCompany(
    companyId: number,
    page = 0,
    size = 20
  ): Promise<AxiosResponse<PaginatedResponse<Job>>> {
    return apiService.get<PaginatedResponse<Job>>(
      `${this.basePath}/company/${companyId}?page=${page}&size=${size}`
    );
  }

  async getJobsByCategory(
    categoryId: number,
    page = 0,
    size = 20
  ): Promise<AxiosResponse<PaginatedResponse<Job>>> {
    return apiService.get<PaginatedResponse<Job>>(
      `${this.basePath}/category/${categoryId}?page=${page}&size=${size}`
    );
  }

  async getJobsBySkills(
    skillIds: number[],
    page = 0,
    size = 20
  ): Promise<AxiosResponse<PaginatedResponse<Job>>> {
    const skillParams = skillIds.map(id => `skillIds=${id}`).join('&');
    return apiService.get<PaginatedResponse<Job>>(
      `${this.basePath}/skills?${skillParams}&page=${page}&size=${size}`
    );
  }

  async getRecommendedJobs(page = 0, size = 20): Promise<AxiosResponse<PaginatedResponse<Job>>> {
    return apiService.get<PaginatedResponse<Job>>(
      `${this.basePath}/recommendations?page=${page}&size=${size}`
    );
  }

  async createJob(jobData: any): Promise<AxiosResponse<Job>> {
    return apiService.post<Job>(this.basePath, jobData);
  }

  async updateJob(id: number, jobData: any): Promise<AxiosResponse<Job>> {
    return apiService.put<Job>(`${this.basePath}/${id}`, jobData);
  }

  async deactivateJob(id: number): Promise<AxiosResponse<void>> {
    return apiService.put<void>(`${this.basePath}/${id}/deactivate`);
  }

  async activateJob(id: number): Promise<AxiosResponse<void>> {
    return apiService.put<void>(`${this.basePath}/${id}/activate`);
  }

  async deleteJob(id: number): Promise<AxiosResponse<void>> {
    return apiService.delete<void>(`${this.basePath}/${id}`);
  }
}

export const jobService = new JobService();
